### PR TITLE
Feature/cb2 6138

### PR DESCRIPTION
**Pull request by bristol-mike-jones**
## Description

This is the last part of paying down the debt from creating these tables manually at the start. The key WMS tables (site, staff and site events) are now managed by Liquibase in this repo.  When an environment is stood up, these tables will get created in the environments DB.

The table creation scripts are based on what it in production, as created by DMS, rather than the scripts that were created manually.  It doesn't seem that the differences are significant, but it means that lower environments match production. The tables are all `create if not exists` so the existing objects will be left as is unless they are cleared down and recreated.

The scripts used to populate data have been collated in this repo.  They now allow either a few days or months worth of site events to be created from the day of being run onwards.  These need to be run manually as its a pain to get these into liquibase, and its not what its for really. 

Related issue: [CB2-6138](https://dvsa.atlassian.net/browse/CB2-6138)

## Before submitting (or marking as "ready for review")

- [ ] Does the pull request title follow the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification?
- [ ] Have you performed a self-review of the code
- [ ] Have you have added tests that prove the fix or feature is effective and working
- [ ] Did you make sure to update any documentation relating to this change?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,35 @@
...
+installed MySQL Database Software
+installed MySQL Connector/J (https://dev.mysql.com/doc/connectors/en/connector-j-binary-installation.html)
+
+## Liquibase Steps
+
+### Create configuration file liquibase.properties, provide database name (default: WMS_EDH_STAGING), database username and password and class path to MySQL J/Connector.
+
+```properties
+driver: com.mysql.cj.jdbc.Driver
+url: jdbc:mysql://localhost:3306/CVSBNOP?createDatabaseIfNotExist=true
```
Comment (ChrisPeska-BJSS) [inline README.md:16]: The example refers to `CVSNOP`, potentially needs to be corrected to `EDH_WMS_STAGING` to better reflect the use case.

File: README.md [hunk 2]
```diff
@@ -1 +1,35 @@
...
+username: root
+password: password
+classpath: mysql-connector-java-8.0.23.jar
```
Comment (ChrisPeska-BJSS) [inline README.md:19]: I would possibly remove any reference to the Connector/J version, the latest one is .31.

File: README.md [hunk 3]
```diff
@@ -1 +1,35 @@
...
+```
+
+### Once database is up and running (database user needs to have privileges in order to create database objects)
+
+Run: `liquibase --defaultsFile liquibase.properties --changeLogFile changelog-master.xml update`
+
+Running without configuration file (provide missing paths / user credentials):
+
+`liquibase --url=jdbc:mysql://localhost:3306/CVSB19155?createDatabaseIfNotExist=true --driver=com.mysql.cj.jdbc.Driver --classpath=mysql-connector-java-8.0.23.jar --username=root --password=password --changeLogFile changelog-master.xml update`
```
Comment (ChrisPeska-BJSS) [inline README.md:28]: It would be good to update the database name in the example.

File: changelog-master.xml [hunk 4]
```diff
@@ -5,4 +5,5 @@
         xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
                       http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
 <include  file="sql/views.sql" relativeToChangelogFile="true"/>
+<include  file="sql/tables.sql" relativeToChangelogFile="true"/>
```
Comment (ChrisPeska-BJSS) [inline changelog-master.xml:8]: Entries in the change-log run in a declared order. If the views depend on the tables created after this is going to fail. Additionally, there is a practice to name them with sequenced prefix 001_ 002_ etc.

File: sql/tables.sql [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+     `created`           DATETIME DEFAULT NULL,
+     `created_by`        DECIMAL(10, 0) NOT NULL,
+     PRIMARY KEY (`id`),
+     KEY nse_site_event_ind1 (`site_event_id`),
+     KEY nse_site_ind2 (`site_id`),
+     KEY nse_staff_ind3 (`staff_id`),
+     KEY nse_event_date_ind4 (`event_date`)
+  )
+  ENGINE = InnoDB;
+
```
Comment (ChrisPeska-BJSS) [inline sql/tables.sql:113]: Throughout all table definitions table names are lower case. CVS Staging database runs with `lower_case_table_names` set to `0` that means name comparisons are case-sensitive. DMS creates tables upper-case in preprod/prod - I assume tables are referred in a similar manner in Lambdas (VTM/WMS).

**Comment by ChrisPeska-BJSS** [comment]
Hi @bristol-mike-jones I can see you've rolled up your sleeves - good starting point to get WMS sorted out. There are some `breaking` problems, easy to fix.

File: sql/tables.sql (hunk 5 above)
Comment (reisedawson-bjss) [reply sql/tables.sql:113]: The table and field names are also all capitalised in dev/int (see queries to generate test data for examples). Great spot Chris.

**Review by reisedawson — COMMENTED** [review]
Looks good Mike, great to get this into the repo! Just Chris' points about capitalisation and removal of CVSNOP refs that I would highlight too, then I'll be happy to approve. 

Can't comment on the liquibase issue as I don't know how it works.

File: changelog-master.xml (hunk 4 above)
Comment (bristol-mike-jones) [reply changelog-master.xml:8]: Ah yes.  I think this is horrid.  The views.sql is hardcoded to a specific schema and has already been deployed to some environments.  MySQL doesn't seem to support the forcible creation of views, so I'm not sure of the best way forward.

I could change the views.sql to be blank now - and then have 002_tables.sql and 003_views.sql?  then on a fresh deployment, the existing view.sql will be blank and so work. for an existing deployment, it would recreate the view (which I think is fine)

Is it safe to take the hardcoded schema out of the view definition? I tried it locally and it seemed fine...

File: sql/tables.sql (hunk 5 above)
Comment (bristol-mike-jones) [reply sql/tables.sql:113]: Will update so that everything is uppercase.  

The reversed table creation scripts I got from prod had everything lowercase, but that's probably the IDE doing it.

File: README.md (hunk 2 above)
Comment (bristol-mike-jones) [reply README.md:19]: I'll make it generic, that its the full path to where ever the jar files lives locally.  I ran it locally and it was happy with the jar file being in my downloads folder - so long as I gave the full path to it

**Comment by bristol-mike-jones** [comment]
@ChrisPeska-BJSS & @reisedawson-bjss - Are you able to take a another look please? I've tried to take you comments into account, but I'm not completely happy with the changes I have had to make around the view definition

**Review by reisedawson — APPROVED** [review]
Looks sensible to me. Is "sql/views.sql" required for something? If it needs to be in the changelog for something I can't think of a cleaner way to do what you've already attempted here so 👍.

**Comment by bristol-mike-jones** [comment]
@ChrisPeska-BJSS  - Are you able to review this PR, please? I would like to get your eyes as well as it was non-trivial with some of the changes.

File: sql/003_views.sql [hunk 6]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (ChrisPeska-BJSS) [inline sql/003_views.sql:1]: Nice trick with moving Views, could you add liquibase metadata at the header? If you add `runOnChange:true` clause, Liquibase will re-run view if the file changes, does not have to be this way.

File: sql/views.sql [hunk 7]
```diff
@@ -1,25 +1,3 @@
 /*
```
Comment (ChrisPeska-BJSS) [inline sql/views.sql:1]: Is this file still required? I looked up `EDH_WMS_STAGING.DATABASECHANGELOG` in Develop and `sql/views.sql` is registered under differed MD5 Hash. This file is going to fail Liquibase once run in Develop.

**Review by ChrisPeska-BJSS — COMMENTED** [review]
It does look neat, there are two niggles, or just one but I am might be missing the bigger Devops-y picture.

File: sql/003_views.sql (hunk 6 above)
Comment (bristol-mike-jones) [reply sql/003_views.sql:1]: I can't get this to work consistently, so I haven't added it.  I have added a comment as to whys the difference between this version of the view and the `views.sql` version

File: sql/views.sql (hunk 7 above)
Comment (bristol-mike-jones) [reply sql/views.sql]: Ah yes, I can't touch the file without Liquibase going mad.  I tried all sorts but it wouldn't work for both a fresh install and an upgrade.  I've reverted the whole file.  On a fresh install, it creates the old view definition and then replaces it with the new one.

File: sql/003_views.sql (hunk 6 above)
Comment (ChrisPeska-BJSS) [reply sql/003_views.sql:1]: I did work for me on Liquibase 4.13 (bit aged version). 
```
--liquibase formatted sql
--changeset liquibase:create -multiple-tables:1 splitStatements:true endDelimiter:; runOnChange:true
```
with the header in the file, any change to the file triggers re-run.

Comment (bristol-mike-jones) [reply sql/003_views.sql:3]: Ah! I had it below the comments....

**Comment by bristol-mike-jones** [comment]
@ChrisPeska-BJSS - Could you re-review, please? I've tested locally on a fresh install and an upgrade from the current develop position.  I've also checked it as a fresh install in the pipeline. Open to other things I can check.

File: sql/003_views.sql [hunk 8]
```diff
@@ -1,10 +1,12 @@
+--liquibase formatted sql
+--changeset liquibase:create -multiple-tables:1 splitStatements:true endDelimiter:; runOnChange:true
 /*
     This view retrieves all the latest events for a booking (DELETED: 0) where a VSA has been allocated
     or the ATF has cancelled the event for billing purposes.
 
     This version of the view does not hardcode the schema to use
 */
-CREATE OR REPLACE VIEW daily_booking_alterations AS
+CREATE OR REPLACE VIEW aa_daily_booking_alterations AS
```
Comment (ChrisPeska-BJSS) [inline sql/003_views.sql:9]: Is the view name change is intended? Or, snacked in accidentally?

File: changelog-master.xml [hunk 9]
```diff
@@ -5,4 +5,6 @@
         xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
                       http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
 <include  file="sql/views.sql" relativeToChangelogFile="true"/>
```
Comment (ChrisPeska-BJSS) [inline changelog-master.xml:7]: Hey, I am not it is painful, do we need it at all? `Develop` has one entry in `EDH_WMS_STAGING.DATABASECHANGELOG` I could just delete it so Liquibase would not complain - `003_views.sql` supersedes it anyway. I could do it in `develop`, I've checked `integration` and my privileges got messed up, but someone with privs could do it.

Comment (bristol-mike-jones) [reply changelog-master.xml:7]: I didn't want to go messing with the liquibase table. Felt like messing with the internals of the tool.  Then there has to be manual steps in the RN and it makes all that more complex and likely to go wrong.

Comment (ChrisPeska-BJSS) [reply changelog-master.xml:7]: `DATABASECHANGELOG` is a registry Liquibase relies on but it is based on a simple concept, where each changed is registered against MD5 hash code, with relevant for CI/CD (not in our case) context information. I would agree that tinkering is not good idea, in this case there is only one entry and the schema had not been managed with Liquibase from the day. I am not going to insist but we could start with cleaner platform by doing so.

**Review by ChrisPeska-BJSS — APPROVED** [review]
👍 from me, @bristol-mike-jones  if you fancy a clean-up-first route give me a shout.
